library/std/src/os/windows/process.rs

```diff
@@ -240,7 +240,8 @@ pub trait CommandExt: Sealed {
     /// This method allows you to specify custom attributes for a child process on Windows systems using raw attribute values.
     /// Raw attributes provide extended configurability for process creation, but their usage can be complex and potentially unsafe.
     ///
-    /// The `attribute` parameter specifies the raw attribute to be set, while the `value` parameter holds the value associated with that attribute.
+    /// The `attribute` parameter specifies the raw attribute to be set, while the `value_ptr` parameter holds the pointer to the value associated with that attribute,
+    /// and the `value_size` parameter indicates the size of the value in bytes.
```

Comment:
I am confused about the required lifetime for `value_ptr`.

In the large example code block at the bottom of https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-updateprocthreadattribute#remarks, it demonstrates a correct `UpdateProcThreadAttribute` call. Roughly:

```c
DWORD ProtectionLevel = ...;

UpdateProcThreadAttribute(...,
                          &ProtectionLevel,
                          sizeof(ProtectionLevel))

CreateProcessW(...)
```

The documentation for `lpValue` says (emphasis in the original):

> A pointer to the attribute value. **This value must persist until the attribute list is destroyed using the DeleteProcThreadAttributeList function.**

Where does the `DeleteProcThreadAttributeList` call occur? Such a thing is not visible in the example, and it's not mentioned anywhere else on that page.

Jumping to conclusions, I assume that `CreateProcessW` calls `DeleteProcThreadAttributeList`, i.e. `*lpValue` (not just `**lpValue`) needs to remain usable not just for the duration of the `UpdateProcThreadAttribute` call, but also after it.

If I'm right about that, I think this needs to be made **a lot more clear** in this Rust documentation.

It wasn't an issue before because the standard library itself would heap-allocate your value for you and keep it until after `spawn` or whatever else does the `CreateProcessW` / `DeleteProcThreadAttributeList`.

If it's now the caller's responsibility, that needs to be a lot more clear with a better example and a discussion of the pitfall of doing `.raw_attribute(..., &parent.as_raw_handle() as *const _);` with a temporary.

Reply:
Here's the revised version with improved grammar and clarity:

> Where does the DeleteProcThreadAttributeList call occur? It's not visible in the example, nor is it mentioned anywhere else on that page.

It is currently invoked as soon as the `ProcThreadAttributeList` is [dropped](https://github.com/rust-lang/rust/pull/121951/files/2a203593551b93dd92858aa65ade20f15ff12c88#diff-50af324626f20bf2db65b4fe30fd5d7b386f5910b5593a002c37ecfa86425df0L887) at the end of the spawn call.

But you bring up a valid point that was also discussed in the original PR.

> Jumping to conclusions, I assume that CreateProcessW calls DeleteProcThreadAttributeList.

Actually, no. You can use the same `ProcThreadAttributeList` to spawn multiple processes. Perhaps it would be better to offload the entire struct and pass a reference to it during the spawn call, something like `spawn_with_attributes`. This way, we could tie a lifetime to the new struct with the minimum lifetime passed through the `raw_attribute` interface.

I will try to write something up.

Reply:
I in vision something like this:

```rust
struct ProcThreadAttributeListBuilder<'a> {
    values: Vec<(*const c_void, usize)>,
    _lifetime_marker: PhantomData<fn() -> &'a ()>,
}

impl<'a> ProcThreadAttributeListBuilder<'a> {
    fn new<T>(a: &'a T) -> Self {
        ProcThreadAttributeListBuilder {
            values: vec![(addr_of!(*a).cast::<c_void>(), size_of::<T>())],
            _lifetime_marker: PhantomData,
        }
    }

    fn attribute<T>(&mut self, value: &'a T) {
        unsafe {
            self.raw_attribute(addr_of!(*value).cast::<c_void>(), size_of::<T>());
        }
    }

    unsafe fn raw_attribute<T>(&mut self, value_ptr: *const T, value_size: usize) {
        self.values.push((value_ptr.cast::<c_void>(), value_size));
    }
}
```

Reply:
I think we need a builder pattern here because the length of the attribute list is immutable as soon as it is created. We might be able to reallocate it and then extend it as soon as the attribute count extends the allocated but I don't know if windows would like that.